traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go

```diff
@@ -263,7 +243,409 @@ func (dss *TODeliveryServiceServer) Delete(db *sqlx.DB, user auth.CurrentUser) (
+				return
+			case tc.DataMissingError:
+				handleErrs(http.StatusBadRequest, err)
+				return
+			case tc.ForbiddenError:
+				handleErrs(http.StatusForbidden, err)
+				return
+			}
+		}
+
+		row := db.QueryRow(selectDeliveryService(), xmlId)
+		var dsId int
+		row.Scan(&dsId)
+
+		// get list of server Ids to insert
+		defer r.Body.Close()
+		payload := GetServersRef()
+
+		if err := json.NewDecoder(r.Body).Decode(payload); err != nil {
+			log.Errorf("Error trying to decode the request body: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		payload.XmlId = xmlId
+		serverNames := payload.ServerNames
+		q, arg, err := sqlx.In(selectServerIds(), serverNames)
+
+		if err != nil {
+			log.Error.Printf("Could not form IN query : %v", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+		q = sqlx.Rebind(sqlx.DOLLAR, q)
+		serverIds, err := db.Query(q, arg...)
+		defer serverIds.Close()
+		if err != nil {
+			log.Error.Printf("Could not select the ServerIds: %v", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// perform the insert transaction
+		rollbackTransaction := true
+		tx, err := db.Beginx()
+		defer func() {
+			if tx == nil || !rollbackTransaction {
+				return
+			}
+			err := tx.Rollback()
+			if err != nil {
+				log.Errorln(errors.New("rolling back transaction: " + err.Error()))
+			}
+		}()
+
+		if err != nil {
+			log.Error.Printf("could not begin transaction: %v", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// We have to get the server Ids and iterate through them because of a bug in the Go
+		// transaction which returns an error if you perform a Select after an Insert in
+		// the same transaction
+		for serverIds.Next() {
+			var serverId int
+			err := serverIds.Scan(&serverId)
+			dtos := map[string]interface{}{"id":dsId, "server":serverId}
+			resultRows, err := tx.NamedQuery(insertIdsQuery(), dtos)
+			if err != nil {
+				if pqErr, ok := err.(*pq.Error); ok {
+					err, eType := dbhelpers.ParsePQUniqueConstraintError(pqErr)
+					log.Error.Printf("could not begin transaction: %v", err)
+					if eType == tc.DataConflictError {
+						handleErrs(http.StatusInternalServerError, err)
+						return
+					}
+					handleErrs(http.StatusInternalServerError, err)
+					return
+				}
+				log.Errorf("received non pq error: %++v from create execution", err)
+				return
+			}
+			resultRows.Next()
+		}
+
+		err = tx.Commit()
+		if err != nil {
+			log.Errorln("Could not commit transaction: ", err)
+			return
+		}
+		rollbackTransaction = false
+
+		// marshal the results to the response stream
+		tcPayload := tc.DeliveryServiceServers{payload.ServerNames, payload.XmlId}
+		payloadResp := tc.DSServersResponse{tcPayload}
+		respBts, err := json.Marshal(payloadResp)
+		if err != nil {
+			log.Errorln("Could not marshal the response as expected: ", err)
+			handleErrs(http.StatusInternalServerError, err)
```

Comment:
`rollbackTransaction = false` needs to come after this. Else, failure to marshal will return an error, but still commit the transaction.

Reply:
in #2282 
NOTE: This is consistent with the CRUD framework. The Marshal function is always called after the transaction has been closed.